File: limbo/library-support/src/main/java/co/windly/limbo/fragment/base/LimboFragmentView.java
package co.windly.limbo.fragment.base;

import android.os.Bundle;
import android.view.View;
import co.windly.limbo.activity.fragment.LimboFragmentActivityView;
import com.hannesdorfmann.mosby3.mvp.MvpView;
import me.yokeyword.fragmentation.ISupportFragment;

public interface LimboFragmentView extends MvpView, ISupportFragment {

  //region Soft Input

  void hideSoftInput();

  void showSoftInput(final View view);

  //endregion

  //region Fragment

  void loadRootFragment(int containerId, LimboFragmentView toFragment);

  void loadRootFragment(int containerId, LimboFragmentView toFragment, boolean addToBackStack, boolean allowAnimation);

  void loadMultipleRootFragment(int containerId, int showPosition, LimboFragmentView... toFragments);

  void start(LimboFragmentView toFragment);

  void start(final LimboFragmentView toFragment, @LimboFragmentView.LaunchMode int launchMode);

  void startForResult(LimboFragmentView toFragment, int requestCode);

  void setFragmentResult(int resultCode, Bundle bundle);

  void startWithPop(LimboFragmentView toFragment);

  void startWithPopTo(LimboFragmentView toFragment, Class<?> targetFragmentClass, boolean includeTargetFragment);

  void replaceFragment(LimboFragmentView toFragment, boolean addToBackStack);

  void showHideFragment(LimboFragmentView showFragment);

  void showHideFragment(LimboFragmentView showFragment, LimboFragmentView hideFragment);

  void pop();

  void popTo(Class<?> targetFragmentClass, boolean includeTargetFragment);

  <T extends LimboFragmentView> T findChildFragment(Class<T> fragmentClass);

  @SuppressWarnings("unchecked")
  <T extends LimboFragmentView> T getParentFragment(Class<T> fragmentClass);

  //endregion

  //region Activity

  <T extends LimboFragmentActivityView> T getActivity(Class<T> activityClass);

  //endregion
}

File: limbo/library/src/main/java/co/windly/limbo/fragment/base/LimboFragmentView.kt
package co.windly.limbo.fragment.base

import android.os.Bundle
import android.view.View
import co.windly.limbo.LimboView
import co.windly.limbo.activity.fragment.LimboFragmentActivityView
import me.yokeyword.fragmentation.ISupportFragment
import me.yokeyword.fragmentation.ISupportFragment.LaunchMode

interface LimboFragmentView : LimboView, ISupportFragment {

  //region Soft Input

  fun hideSoftInput()

  fun showSoftInput(view: View)

  //endregion

  //region Fragment

  fun loadRootFragment(containerId: Int, toFragment: LimboFragmentView)

  fun loadRootFragment(containerId: Int, toFragment: LimboFragmentView, addToBackStack: Boolean,
    allowAnimation: Boolean)

  fun loadMultipleRootFragment(containerId: Int, showPosition: Int, vararg toFragments: LimboFragmentView)

  fun start(toFragment: LimboFragmentView)

  fun start(toFragment: LimboFragmentView, @LaunchMode launchMode: Int)

  fun startForResult(toFragment: LimboFragmentView, requestCode: Int)

  override fun setFragmentResult(resultCode: Int, bundle: Bundle)

  fun startWithPop(toFragment: LimboFragmentView)

  fun startWithPopTo(toFragment: LimboFragmentView, targetFragmentClass: Class<*>, includeTargetFragment: Boolean)

  fun replaceFragment(toFragment: LimboFragmentView, addToBackStack: Boolean)

  fun showHideFragment(showFragment: LimboFragmentView)

  fun showHideFragment(showFragment: LimboFragmentView, hideFragment: LimboFragmentView)

  fun pop()

  fun popTo(targetFragmentClass: Class<*>, includeTargetFragment: Boolean)

  fun <T : LimboFragmentView> findChildFragment(fragmentClass: Class<T>): T?

  fun <T : LimboFragmentView> getParentFragment(fragmentClass: Class<T>): T

  //endregion

  //region Activity

  fun <T : LimboFragmentActivityView> getActivity(activityClass: Class<T>): T

  //endregion
}
